CheckRightsService_GetRightsList

lalagvana opened the pull request:
Some words about the case. We want to get rights list - that means the list of all rights, their names and their description mostly. For example, it'll be needed to see which rights can be chosen by administrator to be given to any user, etc.

- Our endpoint will be like: /GetRightsList. No arguments needed. We just want to see all rights and their attributes (Collection of rights). Idk if it's OK to return Collection...........
- In this case, we don't need any validation actions in GetRightsListRequestValidator. Please, tell me if this file is extra. Or tell me why it's not.
- On repository layer there is just function that requests to dbContext - pls give me Rights Collection!
- In GetRightsListCommand all actions located. Validate->Request to repository->Get holy Rights. Elementary. But is validation should be there at all? Moot question.
- Don't forget about the controller. Routing is located right there.

Also, some words about the Right entity.
- I added Name and Description fields to it. It's connected with our previous discussion: our Rights in CheckRightsService will just have an info about general permissions like "this right gives you an ability to edit news" or "this right gives you an ability to add new members to project crews". No relations with specific users or projects.
- However, I hadn't deleted extra fields. Sure, we have to discuss it again and make this logic more concrete to make some changes and deletions like that.
------
There'll be Unit tests soon. But now I still need your review to resolve mistakes :)
Sure, there are lots of them, hah.

UPD: All my moot questions have been answered. Extra classes deleted. Unit-tests added. Here a bit understanding comes.

lalagvana committed:
Migrations moved into Database directory

lalagvana committed:
Revert "Migrations moved into Database directory"

This reverts commit 91de57bbd916a04ed05f46c513a95b026b3b1679.

lalagvana committed:
Renew entities and moved Migrations

Added new fields to Right (Name and Description) entity but hadn't deleted extra fields (we discussed about new version of these service entities).
Sure, we have to discuss it again and make this logic more concrete to make some changes and deletions.
Also, Migrations moved to Database directory.

lalagvana committed:
Revert "Renew entities and moved Migrations"

This reverts commit 5c526029a64ad1cc52391bfb6df072e4ebc1aea9.

lalagvana committed:
New fields in Right entity, Migrations moved to Database

Added new fields to Right (Name and Description) entity but hadn't deleted extra fields (we discussed about new version of these service entities).
Sure, we have to discuss it again and make this logic more concrete to make some changes and deletions.
Also, Migrations moved to Database directory.

lalagvana committed:
Empty test files

lalagvana committed:
Something unnecessary

lalagvana committed:
Created directories and empty classes

lalagvana committed:
Commented field UserId

Trying to exclude some extra fields as it can be excluded in such structure.

lalagvana committed:
Realization of getRightsList endpoint

*Our endpoint will be like: /GetRightsList. No arguments needed. We just want to see all rights and their atributes(Collection of rights). Idk if it's OK to return Collection...........
*In this case, we don't need any validation actions in GetRightsListRequestValidator. Please, tell me if this file is extra. Or tell me why it's not.
*On repository layer there is just function that requests to dbContext - pls give me Rights Collection!
*In GetRightsListCommand all actions located.
Validate->Request to repository->Get holy Rigths. Elementarily.
*Don't forget about the controller. Routing is located right there.

DarkFelas commented on DigitalOffice/Services/CheckRightsService/Commands/GetRightsListCommand.cs:
```diff
@@ -0,0 +1,35 @@
+{
+    public class GetRightsListCommand : IGetRightsListCommand
+    {
+        private readonly IValidator<GetRightsListRequest> validator;
```
No need, remove

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Services/CheckRightsService/Commands/GetRightsListCommand.cs:
```diff
@@ -0,0 +1,35 @@
+        private readonly IValidator<GetRightsListRequest> validator;
+        public GetRightsListCommand([FromServices] IValidator<GetRightsListRequest> validator)
+        {
+            this.validator = validator;
```
remove

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Services/CheckRightsService/Commands/GetRightsListCommand.cs:
```diff
@@ -0,0 +1,35 @@
+        }
+        public ICollection<Right> Execute(GetRightsListRequest request)
+        {
+            try
```
remove

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Services/CheckRightsService/Controllers/CheckRightsController.cs:
```diff
@@ -0,0 +1,22 @@
+    public class CheckRightsController : ControllerBase
+    {
+        [Route("GetRightsList")]
+        [HttpGet]  // Get /GetRightsList
```
remove comment

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Services/CheckRightsService/Controllers/CheckRightsController.cs:
```diff
@@ -0,0 +1,22 @@
+        [HttpGet]  // Get /GetRightsList
+        public ICollection<Right> GetRightsByUserId(
+            [FromServices] IGetRightsListCommand command,
+            [FromServices] GetRightsListRequest request)
```
request ???? what for?

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Services/CheckRightsService/Database/Entities/Right.cs:
```diff
@@ -9,7 +9,9 @@ public class Right
+        public string Name { get; set; }
+        public string Description { get; set; }
+      //  public Guid UserId { get; set; }
         public ICollection<RightTypeLink> Types { get; set; }
```
Remove

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Services/CheckRightsService/Database/Entities/Right.cs:
```diff
@@ -9,7 +9,9 @@ public class Right
+        public string Description { get; set; }
+      //  public Guid UserId { get; set; }
         public ICollection<RightTypeLink> Types { get; set; }
         public ICollection<RightProjectLink> PermissionsIds { get; set; }
```
Remove

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Services/CheckRightsService/Commands/Interfaces/IGetRightsListCommand.cs:
```diff
@@ -0,0 +1,11 @@
+{
+    public interface IGetRightsListCommand
+    {
+        ICollection<Right> Execute(GetRightsListRequest request);
```
Remove request

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Services/CheckRightsService/Commands/GetRightsListCommand.cs:
```diff
@@ -0,0 +1,35 @@
+            {
+                throw new Exception("Error.");
+            }
+            var repository = new CheckRightsRepository(new CheckRightsServiceDbContext(new DbContextOptions<CheckRightsServiceDbContext>()));
```
use DI for repository

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Services/CheckRightsService/Repositories/CheckRightsRepository.cs:
```diff
@@ -0,0 +1,22 @@
+        }
+        public ICollection<Right> GetRightsList()
+        {
+            ICollection<Right> rights = (ICollection<Right>)dbContext.Rights;
```
Do not send db model back, create new DTO model Right. Rename db model -> DbRight, Create mapper and map DbRight -> Right

lalagvana replied:
Does new directory Models have to be created? Or new DTO model and db model can both be in Entities directory?

lalagvana replied:
fixed

lalagvana committed:
Test preset

DarkFelas commented on DigitalOffice/Services/CheckRightsService/Repositories/Interfaces/ICheckRightsRepository.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    public interface ICheckRightsRepository
+    {
+        ICollection<Right> GetRightsList();
```
IEnumerable

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Services/CheckRightsService/RestRequest/GetRightsListRequest.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace CheckRightsService.RestRequest
+{
+    public class GetRightsListRequest
```
Remove

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Services/CheckRightsService/Validators/GetRightsListRequestValidator.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace CheckRightsService.Validators
+{
+    public class GetRightsListRequestValidator : AbstractValidator<GetRightsListRequest>
```
Remove

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Tests/CheckRightsServiceTests/CommandsTests/GetRightsListCommandTests.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace CheckRightsServiceUnitTests.CommandsTests
+{
+    class GetRightsListCommandTests
```
Empty?

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Tests/CheckRightsServiceTests/RepositoriesTests/CheckRightsRepositoryTests.cs:
```diff
@@ -0,0 +1,10 @@
+using System.Text;
+
+namespace CheckRightsServiceUnitTests.RepositoriesTests
+{
```
Empty?

DarkFelas commented on DigitalOffice/Tests/CheckRightsServiceTests/ValidatorsTests/GetRightsListRequestValidator.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace CheckRightsServiceUnitTests.ValidatorsTests
+{
+    class GetRightsListRequestValidator
```
Remove

lalagvana replied:
fixed

lalagvana committed:
Deleted extra classes etc.

lalagvana committed:
Extra usings deleted

lalagvana committed:
Deleted extra entities that causes errors

lalagvana committed:
Revert "Deleted extra entities that causes errors"

This reverts commit 320f99433b707784af16e0952195c69052e2ad91.

lalagvana committed:
Deleted extra entities that generate errors

lalagvana committed:
Update DbContext

lalagvana committed:
IEnumerable instead of ICollection and changes in repository initialization

DarkFelas commented on DigitalOffice/Services/CheckRightsService/Commands/GetRightsListCommand.cs:
```diff
@@ -0,0 +1,18 @@
+        private readonly CheckRightsServiceDbContext dbContext;
+        public IEnumerable<Right> Execute()
+        {
+            var repository = new CheckRightsRepository(dbContext);
```
Use DI not new()

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Services/CheckRightsService/Repositories/CheckRightsRepository.cs:
```diff
@@ -0,0 +1,22 @@
+        public CheckRightsRepository(CheckRightsServiceDbContext dbContext)
+        {
+            this.dbContext = dbContext;
+        }
```
empty line after brace

DarkFelas commented on DigitalOffice/Services/CheckRightsService/Commands/GetRightsListCommand.cs:
```diff
@@ -0,0 +1,18 @@
+{
+    public class GetRightsListCommand : IGetRightsListCommand
+    {
+        private readonly CheckRightsServiceDbContext dbContext;
```
empty line after between field and method

lalagvana replied:
fixed

lalagvana committed:
Mistakes were resolved, tests were added

-Empty lines added
-Resolved mistakes with "new repository" and repository interface
-Added missed .AddTransient in Startup.cs
-Moq installed
-Added tests for Command and Repository

lalagvana committed:
DbRight and RightDto

Added and renamed these and their links

lalagvana committed:
Tried to return DTO model

djino97 commented on DigitalOffice/Services/CheckRightsService/Controllers/CheckRightsController.cs:
```diff
@@ -0,0 +1,21 @@
+        public IEnumerable<RightDto> GetRightsList([FromServices] IGetRightsListCommand command)
+        {
+            IEnumerable<RightDto> rights = (IEnumerable<RightDto>) command.Execute();
+            return rights;
```
`return command.Execute()` and controller should return value like  `IEnumerable<DbRight>`.

lalagvana replied:
fixed, controller returns List<Right>

lalagvana committed:
Fixed, mapper to DTO model added

lalagvana committed:
Some hidden errors in tests to test build

lalagvana committed:
Fixed mapper and repository test

lalagvana committed:
Migration added

denis-kriv commented on DigitalOffice/Services/CheckRightsService/Models/Right.cs:
```diff
@@ -0,0 +1,11 @@
+{
+    public class Right
+    {
+        public int Id { get; set; }
```
Why id is int? Maybe it should be guid?

lalagvana replied:
No, this Id is going to be sent to front, but it has another meaning - Number from DbRight model.

lalagvana replied:
Also, about tests. 
Could you please offer some new conditionals that we have to check for these methods?
This is a GET operation, so we don't need any other checks, only "returns list successfully", I suppose. There are no validation or some other conditionals as I see it.

denis-kriv reviewed (commented):
I think we should add more than one test for each method

MihailKulikov commented on DigitalOffice/Tests/CheckRightsServiceTests/CommandsTests/GetRightsListCommandTests.cs:
```diff
@@ -0,0 +1,27 @@
+using Moq;
+using NUnit.Framework;
+
+namespace CheckRightsServiceUnitTests
```
Fix namespace

lalagvana replied:
fixed

lalagvana committed:
Namespace fixed, extra usings deleted

lalagvana commented:
> I think we should add more than one test for each method

Could you please offer some new conditionals that we have to check for these methods?
This is a Get operation, so we don't need any other checks, only "returns list successfully", I suppose. There are no validation or some other conditionals as I see it.

denis-kriv commented:
> I think we should add more than one test for each method
> 
> Could you please offer some new conditionals that we have to check for these methods?
> This is a Get operation, so we don't need any other checks, only "returns list successfully", I suppose. There are no validation or some other conditionals as I see it.

Ok, I can not think any cases

MihailKulikov commented on DigitalOffice/Services/CheckRightsService/Startup.cs:
```diff
@@ -26,6 +34,10 @@ public void ConfigureServices(IServiceCollection services)
                 options.UseSqlServer(Configuration.GetConnectionString("SQLConnectionString"));
             });
+
+            services.AddTransient<ICheckRightsRepository, CheckRightsRepository>();
+            services.AddTransient<IGetRightsListCommand, GetRightsListCommand>();
+            services.AddTransient<IMapper<DbRight, Right>, RightsMapper>();
```
Maybe will be a good idea to create separate method for binding DI container?

lalagvana replied:
Yep, great idea

lalagvana replied:
Done

DarkFelas replied:
Where? I do not see any new methods? Create separated methods for each type of object e.g. ConfigureCommands, ConfigureRepositories and others ...

lalagvana committed:
Method for binding DI container

DarkFelas commented on DigitalOffice/Services/CheckRightsService/CheckRightsService.csproj:
```diff
@@ -1,17 +1,21 @@
+    <PackageReference Include="FluentValidation" Version="9.0.1" />
+    <PackageReference Include="FluentValidator" Version="2.0.4" />
     <PackageReference Include="Microsoft.EntityFrameworkCore" Version="3.1.6" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.InMemory" Version="3.1.6" />
```
Remove

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Services/CheckRightsService/CheckRightsService.csproj:
```diff
@@ -1,17 +1,21 @@
       <PrivateAssets>all</PrivateAssets>
       <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
     </PackageReference>
+    <PackageReference Include="Moq" Version="4.14.5" />
```
REmove

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Services/CheckRightsService/Database/Entities/DbRight.cs:
```diff
@@ -0,0 +1,15 @@
+    public class DbRight
+    {
+        [Key]
+        public Guid Id { get; set; }
```
int

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Services/CheckRightsService/Database/Entities/DbRight.cs:
```diff
@@ -0,0 +1,15 @@
+    {
+        [Key]
+        public Guid Id { get; set; }
+        public int Number { get; set; }
```
Remove

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Services/CheckRightsService/Mappers/RightsMapper.cs:
```diff
@@ -0,0 +1,19 @@
+        {
+            return new Right
+            {
+                Id = value.Number,
```
Rework

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Services/CheckRightsService/Migrations/20200724101649_InitialMigration.cs:
```diff
@@ -0,0 +1,31 @@
+                columns: table => new
+                {
+                    Id = table.Column<Guid>(nullable: false),
+                    Number = table.Column<int>(nullable: false),
```
REmove

lalagvana replied:
fixed

DarkFelas replied:
I still see Number here and Id is Guid

DarkFelas commented on DigitalOffice/Services/CheckRightsService/Startup.cs:
```diff
@@ -26,6 +34,15 @@ public void ConfigureServices(IServiceCollection services)
+            AddTransientObjects(services);
+        }
+
+        public void AddTransientObjects(IServiceCollection services)
```
Not good

lalagvana replied:
fixed

lalagvana committed:
Some fixes

DarkFelas commented on DigitalOffice/Services/CheckRightsService/CheckRightsService.csproj:
```diff
@@ -1,11 +1,13 @@
 
   <ItemGroup>
+    <PackageReference Include="FluentValidation" Version="9.0.1" />
+    <PackageReference Include="FluentValidator" Version="2.0.4" />
```
Remove

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Tests/CheckRightsServiceTests/CommandsTests/GetRightsListCommandTests.cs:
```diff
@@ -0,0 +1,27 @@
+        [Test]
+        public void GetRightsListCommandSuccess()
+        {
+            Assert.DoesNotThrow(() => command.Execute());
```
Add setup to repositoryMock which return something.

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Tests/CheckRightsServiceTests/RepositoriesTests/CheckRightsRepositoryTests.cs:
```diff
@@ -0,0 +1,37 @@
+            dbOptionsGetRightsList = new DbContextOptionsBuilder<CheckRightsServiceDbContext>()
+                                    .UseInMemoryDatabase(databaseName: "InMemoryDatabase")
+                                    .Options;
+            dbContext = new CheckRightsServiceDbContext(dbOptionsGetRightsList);
```
Put something in dbContext and check that repository return something.

lalagvana replied:
fixed

lalagvana committed:
Updated tests

-Added Mappers tests
-Added setup to repository in Command tests
-Added info in dbContext to check if repository return smth

lalagvana committed:
Merge branch 'develop' into feature/CheckRightsService_GetRights

lalagvana committed:
Fixes after merging

Deleted extra classes

DarkFelas commented on DigitalOffice/Services/CheckRightsService/Migrations/CheckRightsServiceDbContextModelSnapshot.cs:
```diff
@@ -19,126 +19,25 @@ protected override void BuildModel(ModelBuilder modelBuilder)
+                        .HasColumnType("nvarchar(max)");
 
-                    b.Property<int>("RightType")
+                    b.Property<int>("Number")
```
Number again

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Services/CheckRightsService/Migrations/CheckRightsServiceDbContextModelSnapshot.cs:
```diff
@@ -19,126 +19,25 @@ protected override void BuildModel(ModelBuilder modelBuilder)
-            modelBuilder.Entity("CheckRightsService.Database.Entities.Right", b =>
+            modelBuilder.Entity("CheckRightsService.Database.Entities.DbRight", b =>
                 {
                     b.Property<Guid>("Id")
```
Guid?

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Tests/CheckRightsServiceTests/CommandsTests/GetRightsListCommandTests.cs:
```diff
@@ -0,0 +1,36 @@
+        {
+            var right = new Right { Id = 0, Name = "Right", Description = "Allows you everything" };
+            var rightsList = new List<Right> {right};
+            repositoryMock.Setup(repository => repository.GetRightsList()).Returns(rightsList);
```
Move to SetUp method

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Tests/CheckRightsServiceTests/RepositoriesTests/CheckRightsRepositoryTests.cs:
```diff
@@ -0,0 +1,49 @@
+                                    .UseInMemoryDatabase(databaseName: "InMemoryDatabase")
+                                    .Options;
+            dbContext = new CheckRightsServiceDbContext(dbOptions);
+            if (dbContext.Database.IsInMemory())
```
read do-backend-wiki

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Tests/CheckRightsServiceTests/RepositoriesTests/CheckRightsRepositoryTests.cs:
```diff
@@ -0,0 +1,49 @@
+        public void GetRightsListSuccessfully()
+        {
+            var dbRight = new DbRight { Id = 0, Name = "Right", Description = "Allows you everything"};
+            dbContext.Rights.Add(dbRight);
```
move to SetUp method

lalagvana replied:
fixed

DarkFelas commented on DigitalOffice/Tests/CheckRightsServiceTests/RepositoriesTests/CheckRightsRepositoryTests.cs:
```diff
@@ -0,0 +1,49 @@
+            var dbRight = new DbRight { Id = 0, Name = "Right", Description = "Allows you everything"};
+            dbContext.Rights.Add(dbRight);
+            dbContext.SaveChanges();
+            mapperMock.Setup(mapper => mapper.Map(dbRight)).Returns(new Right { Id = dbRight.Id, Name = dbRight.Name, Description = dbRight.Description });
```
Move to SetUp method

lalagvana replied:
fixed

lalagvana committed:
Small fixes

-Fixed migration.
-Added separate methods in Startup.cs.
-Moved some things from Test to Startup method.
